src/ctia/task/migration/migrate_es_stores.clj

```diff
@@ -75,6 +75,79 @@
   (let [coercer (coerce-to-fn Model)]
     #(reduce (partial append-coerced coercer) {} %)))
 
+(defn read-source
+  "This function retrieves in `source-store` all the documents that match the given `query`
+   by batch of maximum size `batch-size`. When not nil the `search_after` parameter is used
+   to skip previously retrieved data. Then it pushes the retrieved data in the given channel,
+  `data-chan`, along with next search_after value for eventual further restart. Once all the
+   documents that match given query are retrieved, it closes the channel `data-chan`"
+  [{:keys [source-store
+           search_after
+           data-chan
+           batch-size
+           query]}]
+  (loop [current-search-after search_after]
+    ;; loop that reads batches from source and produces them in chan
+    (let [{:keys [data paging] :as res} (mst/query-fetch-batch query
+                                                               source-store
+                                                               batch-size
+                                                               0
+                                                               "asc"
+                                                               current-search-after)
+          next (:next paging)
+          next-search-after (:sort paging)]
+      (>!! data-chan {:documents data
+                      :search_after next-search-after})
+      (if next
+        (recur  next-search-after)
+        (close! data-chan)))))
+
+(defn write-target
+  "This function reads batches from `data-chan` (pushed by read-source) until it reads
+   a `nil`, meaning that the channel has been closed by `read-source` function. The
+   batches are (1) modified with `migrations` functions, (2) validated by `list-coerce`
+   and (3) written into given `target-store`. At each batch the migration state, number
+   of migrated documents and search_after value, is updated in ES after the documents
+   are written in the store. If the process fails or is stopped before that update of
+   the migration state, the documents will be red again by `read-source` and overridden
+   in case of restart."
+  [{:keys [target-store
+           data-chan
+           migrations
+           entity-type
+           batch-size
+           migration-id
+           list-coerce
+           migrated-count
+           migration-es-conn
+           confirm?]}]
+  (loop [current-count migrated-count]
+    ;; sync loop that consumes batches in chan and migrates them to target
+    (if-let [{:keys [documents search_after]} (<!! data-chan)]
+      (let [migrated (transduce migrations conj documents)
+           {:keys [data errors]} (list-coerce migrated)
+           new-current-count (+ current-count (count data))]
+        (doseq [entity errors]
```

Comment:
I would test the errors list for emptiness first, it feels weird in the code, like you should always have errors